aws_lambda_powertools/utilities/idempotency/persistence/redis.py

```diff
@@ -0,0 +1,440 @@
+            #    - first time that this invocation key is used
+            #    - previous invocation with the same key was deleted due to TTL
+            #    - SET see https://redis.io/commands/set/
+
+            logger.debug(f"Putting record on Redis for idempotency key: {item['name']}")
+            encoded_item = self._json_serializer(item["mapping"])
+            ttl = self._get_expiry_second(expiry_timestamp=item["mapping"][self.expiry_attr])
+
+            redis_response = self.client.set(name=item["name"], value=encoded_item, ex=ttl, nx=True)
+
+            # redis_response:True -> Redis set succeed, idempotency key does not exist before
+            # return to idempotency and proceed to handler execution phase. Most cases should return here
+            if redis_response:
+                return
+
+            # redis_response:None -> Existing record on Redis, continue to checking phase
+            # The idempotency key exist:
+            #    - previous invocation with the same key and not expired(active idempotency)
+            #    - previous invocation timed out (Orphan Record)
+            #    - previous invocation record expired but not deleted by Redis (Orphan Record)
+
+            idempotency_record = self._get_record(item["name"])
+
+            # status is completed and expiry_attr timestamp still larger than current timestamp
+            # found a valid completed record
+            if idempotency_record.status == STATUS_CONSTANTS["COMPLETED"] and not idempotency_record.is_expired:
+                raise IdempotencyItemAlreadyExistsError
+
+            # in_progress_expiry_attr exist means status is in_progress, and still larger than current timestamp,
+            # found a vaild in_progress record
+            if (
+                idempotency_record.status == STATUS_CONSTANTS["INPROGRESS"]
+                and idempotency_record.in_progress_expiry_timestamp
+                and idempotency_record.in_progress_expiry_timestamp > int(now.timestamp() * 1000)
+            ):
+                raise IdempotencyItemAlreadyExistsError
+
+            # If the code reaches here means we found an Orphan record.
+            raise IdempotencyOrphanRecordError
+
+        except IdempotencyOrphanRecordError:
+            # deal with orphan record here
+            # aquire a lock for default 10 seconds
+            with self._acquire_lock(name=item["name"]):
+                self.client.set(name=item["name"], value=encoded_item, ex=ttl)
+
+            # lock was not removed here intentionally. Prevent another orphan operation in race condition.
+        except (redis.exceptions.RedisError, redis.exceptions.RedisClusterException) as e:
+            raise e
+        except Exception as e:
+            logger.debug(f"encountered non-redis exception:{e}")
+            raise e
+
+    @contextmanager
+    def _acquire_lock(self, name: str):
+        """
+        aquire a lock for default 10 seconds
```

Comment:
```suggestion
        Attempt to acquire a lock for a specified resource name, with a default timeout.

        This context manager attempts to set a lock using Redis to prevent concurrent
        access to a resource identified by 'name'. It uses the 'nx' flag to ensure that
        the lock is only set if it does not already exist, thereby enforcing mutual exclusion.
```